packages/graphback-datasync/src/providers/DatasyncMongoDBDataProvider.ts

```diff
@@ -10,8 +9,8 @@ import { DataSyncProvider } from "./DataSyncProvider";
  */
 export class DataSyncMongoDBDataProvider<Type = any> extends MongoDBDataProvider<Type> implements DataSyncProvider {
 
-  public constructor(baseType: GraphQLObjectType, client: any) {
-    super(baseType, client);
+  public constructor(model: ModelDefinition, client: any) {
```

Comment:
Now we are passing the model directly to DataProvider. 

The way I see it in the future https://github.com/aerogear/graphback/issues/1675 is that the DataProvider will have minimal knowledge on how to operate on a GraphQL AST Objects e.g a `GraphQLObjectType` to retrieve indexes etc,because once the model is enriched with appropriate informations e.g fields and their types, relationships, model name, corresponding database name, column mapping (if any), indexes etc there won't be any need to play around with GraphQL Object types to get these information at this stage.

Reply:
See (createIndexes.ts)[https://github.com/aerogear/graphback/blob/2b355e62f73aae04de10878701e4fb595fcfb482/packages/graphback-runtime-mongodb/src/utils/createIndexes.ts] file to see example of such operations.

Reply:
> there won't be any need to play around with GraphQL Object types to get these information at this stage.

Absolutely - in best practices the data providers should not have any reference to a GraphQLObjectTypes at all. By abstracting ModelDefinition completely away from GraphQLObjectType the providers could be used outside of a GraphQL application - which would be awesome.

However, the `ModelDefinition` does still need a reference to the GraphQLObjectType for use elsewhere, such as in GraphbackCoreMetadata when building a plugin.

Perhaps these interfaces should be separate, where the Data Provider accepts a different interface from ModelDefinition containing only data valuable to the dataprovider.

I am a little bit against having an abstraction of GraphQLObjectType that can be used in both places (meaning it would no longer have any reference to GraphQLObjectType) because in older versions of Graphback we had this, called `InputModelTypeContext` and it was difficult to work with. See https://github.com/aerogear/graphback/blob/0.10.1/packages/graphback-core/src/api/ContextTypes.ts#L79-L86

Reply:
> > there won't be any need to play around with GraphQL Object types to get these information at this stage.
> 
> Absolutely - in best practices the data providers should not have any reference to a GraphQLObjectTypes at all. By abstracting ModelDefinition completely away from GraphQLObjectType the providers could be used outside of a GraphQL application - which would be awesome.
> 
> However, the `ModelDefinition` does still need a reference to the GraphQLObjectType for use elsewhere, such as in GraphbackCoreMetadata when building a plugin.
>

Yes, indeed. This has not changed for the moment. 
 
> Perhaps these interfaces should be separate, where the Data Provider accepts a different interface from ModelDefinition containing only data valuable to the dataprovider.
>

We can explore this in separate PR, and preferably in https://github.com/aerogear/graphback/issues/1675. 
 
> I am a little bit against having an abstraction of GraphQLObjectType that can be used in both places (meaning it would no longer have any reference to GraphQLObjectType) because in older versions of Graphback we had this, called `InputModelTypeContext` and it was difficult to work with. See https://github.com/aerogear/graphback/blob/0.10.1/packages/graphback-core/src/api/ContextTypes.ts#L79-L86

We ca examine how the referenced `GraphQLObjectType` is used and see if we can get off more metadata from the GraphQL type to something more usable. This should be part of  https://github.com/aerogear/graphback/issues/1675

Reply:
Yep, I know it is not changing here just adding my thoughts to your idea!